controller/controller_server/utils.py

```diff
@@ -8,43 +10,147 @@
 import controller.controller_server.messages as messages
 from controller.array_action.config import FC_CONNECTIVITY_TYPE, ISCSI_CONNECTIVITY_TYPE
 from controller.common.csi_logger import get_stdout_logger
+from controller.controller_server.controller_types import ArrayConnectionInfo, ObjectIdInfo, ObjectParameters
 from controller.controller_server.errors import ObjectIdError, ValidationException
 from controller.csi_general import csi_pb2
 
 logger = get_stdout_logger()
 
 
-def get_array_connection_info_from_secret(secrets):
+def _is_topology_match(secret_topologies, node_topologies):
+    for topologies in secret_topologies:
+        logger.debug(
+            "Comparing topologies: volume topologies: {},"
+            " node topologies: {}".format(topologies, node_topologies))
+        if topologies.items() <= node_topologies.items():
+            return True
+    return False
+
+
+def get_volume_topologies(request):
+    if request.HasField(config.REQUEST_ACCESSIBILITY_REQUIREMENTS_FIELD):
+        accessibility_requirements = request.accessibility_requirements
+        if accessibility_requirements.preferred:
+            topologies = accessibility_requirements.preferred[0].segments
+            logger.info("Chosen volume topologies: {}".format(topologies))
+            return topologies
+
+
+def get_secret_by_topologies(raw_secret_config, node_topologies):
+    secret_config = ast.literal_eval(raw_secret_config)
+    for system_id, secret_info in secret_config.items():
+        secret_topologies = secret_info.get(config.SECRET_SUPPORTED_TOPOLOGIES_PARAMETER)
+        if _is_topology_match(secret_topologies, node_topologies):
+            return secret_info, system_id
+
+
+def get_secrets_by_system_id(secret_config_string, system_id):
+    secret_config = ast.literal_eval(secret_config_string)
+    return secret_config.get(system_id)
+
+
+def _get_system_info_from_secrets(secrets, topologies=None, system_id=None):
+    secret_config = secrets.get(config.SECRET_CONFIG_PARAMETER)
+    system_secrets = secrets
+    if secret_config:
+        if system_id:
+            system_secrets = get_secrets_by_system_id(secret_config_string=secret_config, system_id=system_id)
+        elif topologies:
+            system_secrets, system_id = get_secret_by_topologies(raw_secret_config=secret_config,
+                                                                 node_topologies=topologies)
```

Comment:
`raw_secret_config` can be parsed right after `if secret_config:`, so none of the other two functions would need to worry about parsing it

Reply:
done